### PR TITLE
Fix replication

### PR DESCRIPTION
**Pull request by cole**
Relates to: https://github.com/archivematica/Issues/issues/684
Relates to: https://github.com/archivematica/Issues/issues/463

### CODE REVIEW

**Comment by replaceafill** [comment]
@cole Do you think your fix also fixes https://github.com/archivematica/Issues/issues/463?

**Comment by cole** [comment]
@replaceafill it seems like it should, I didn't test that case though.

File: storage_service/locations/models/package.py [hunk 1]
```diff
@@ -1183,12 +1192,12 @@ def create_new_pointer_file_with_replication(
         replication_event = self.create_replication_event(
             replica_package, event_uuid=replication_event_uuid, agents=ss_agents
         )
-        old_premis_events.append(replication_event)
+        old_premis_events += (replication_event,)
```
Comment (ross-spencer) [inline storage_service/locations/models/package.py]: I suspect I'm missing something obvious but I'm not sure why this change is needed for this? `old_premis_events` I believe to be a list at this stage so adding a one item tuple, or appending to the list has the same effect? Same q. with the other `+= (tuple, )` lines below.

File: storage_service/locations/tests/test_package.py [hunk 2]
```diff
@@ -357,6 +357,27 @@ def test_run_post_store_callbacks_dip(self):
...
+            relative_path=replication_dir,
+            purpose=models.Location.REPLICATOR,
+        )
+
+        aip.create_replicas()
+        replica = aip.replicated_package
+
+        assert replica is not None
+        # The relationship is a little bit broken here; one would think that the
+        # _original_ AIP should have replicas, and the replicated AIP should not.
```
Comment (ross-spencer) [inline storage_service/locations/tests/test_package.py:377]: I agree! The database is a little cleaner in this regard, and just has a `replicated_package_id` so I guess we just work it out from that.

**Review by ross-spencer — APPROVED** [review]
Just a clarification asked for my own learning, but otherwise looks great. Thanks for the quick fix! 

One point, and this is coming from someone who I think might not be great at this, maybe a note for a potential process in the future, but a diagnosis of the issue, in the PR or the issue itself would be great. I feel like it would speed up CR but also help when it comes to discuss in joint analyst/dev meetings too. 

I think I can see here that I introduced an assumption with https://github.com/artefactual/archivematica-storage-service/commit/3f21a96879103d6e550752770e344cff42cd919a that self.origin_location was initialized, but that isn't the case when creating a replicated package, and so we need to be more careful with that before determining whether a pointer file is created/updated.

**Comment by ross-spencer** [comment]
My test shows archivematica/Issues#463 is fixed as well. A point of testing for the analysts, (based on the chat in Slack between yourself and Evelyn) should be to:

* [ ] create a replicated packaged.
* [ ] ensure the replication entries in the pointer file are correct.

**Comment by cole** [comment]
Thanks for the feedback @ross-spencer! I may have been premature in marking this as ready for CR, as it hasn't yet been tagged with the 1.10 milestone.

File: storage_service/locations/models/package.py (hunk 1 above)
Comment (cole) [reply storage_service/locations/models/package.py]: This is a good point — I was getting a "tuple has no attribute append" here, but looking at the metsrw code again, I don't see it returning a tuple. I'll have another look.

File: storage_service/locations/tests/test_package.py (hunk 2 above)
Comment (cole) [reply storage_service/locations/tests/test_package.py:377]: It looks like `replicated_package_id` is also backwards (I think unintentionally, based on the [docstring of replicate](https://github.com/artefactual/archivematica-storage-service/blob/stable/0.14.x/storage_service/locations/models/package.py#L468)), that is, the original package will have a value for `replicated_package_id`, and the replica package will not. This seems to me to be the cause of https://github.com/artefactual/archivematica-storage-service/issues/270. It needs to be looked at more though.

**Comment by cole** [comment]
> I think I can see here that I introduced an assumption with 3f21a96 that self.origin_location was initialized, but that isn't the case when creating a replicated package, and so we need to be more careful with that before determining whether a pointer file is created/updated.

Yeah, I think so, but I think it was broken prior to that. As soon as I fixed the origin_location issue, I started getting the error detailed in https://github.com/archivematica/Issues/issues/463; I'm not clear if that was caused by a metsrw update such as f70a99c or something else. I don't see any smoking guns in metsrw or how it is used in storage service though.

I guess it's good to be careful, but to me the takeaway is that we need more meaningful validation of changes (i.e. unit tests). It's easy to miss things when developing! There are a lot of possible configurations, to the point where I don't think everyone can remember to account for all of them (especially new-ish contributors who like me aren't even aware of all features).

**Comment by ross-spencer** [comment]
> I guess it's good to be careful, but to me the takeaway is that we need more meaningful validation of changes (i.e. unit tests). It's easy to miss things when developing! There are a lot of possible configurations, to the point where I don't think everyone can remember to account for all of them (especially new-ish contributors who like me aren't even aware of all features).

Totally!

File: storage_service/locations/models/package.py (hunk 1 above)
Comment (cole) [reply storage_service/locations/models/package.py]: I think I got the lines mixed up here: `new_relationships` is a tuple, but `old_premis_events` and `old_premis_agents` are lists. `+=` also works for lists, but I'll switch back to `append`.

**Comment by cole** [comment]
@replaceafill, can you review the unit test I added to cover https://github.com/archivematica/Issues/issues/463?

**Review by replaceafill — COMMENTED** [review]
LGTM! :+1:
